tools/noescape/noescape.go

```diff
@@ -0,0 +1,396 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package noescape allows recursive escape analysis for hot paths.
+//
+// To the use the package, annotation a function-level comment with either the
+// line "// +noescape" or "// +noescape:OPTION[,OPTION]". In the second case,
+// only the given types will be checked. Valid types include:
+//
+// 	interface: A call is made via an interface.
+// 	heap: A direct allocation is made on the heap.
+// 	builtin: A call is made to a built-in allocation function.
+// 	closure: A closure is allocated.
+// 	dynamic: A dynamic function is dispatched (similar to an interface).
+//
+// The key 'local' may also be specified as an OPTION. This means that only
+// local escapes are considered: escapes in subcalls are allowed.
+//
+// The key 'all' may also be specified as an OPTION. This includes all reasons,
+// local to the function and in invoked functions. Similarly, the key
+// 'explicit' may be specified, which limits the scope to hard escapes only.
+//
+// Some examples of this syntax:
+//
+// +noescape:all               - Analyzes for all escapes in this function and all calls.
+// +noescape:local             - Analyzes only for default local escapes.
+// +noescape                   - Does the same as +noescape:local,explicit.
+// +noescape:heap              - Only analyzes for heap escapes.
+// +noescape:interface,dynamic - Only checks for dynamic calls and interface calls.
+//
+// Local examples can be made by a comment of the form "// escapes: reason."
+// This must appear on the line of the escape.
+//
+package noescape
+
+import (
+	"encoding/gob"
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/ssa"
+)
+
+// Analyzer defines the entrypoint.
+var Analyzer = &analysis.Analyzer{
+	Name: "noescape",
+	Doc:  "surfaces recursive escape analysis results",
+	Requires: []*analysis.Analyzer{
+		buildssa.Analyzer,
+	},
+	Run: run,
+}
+
+// packageEscapeFacts is the set of all functions in a package, and whether or
+// not they recursively pass escape analysis.
+//
+// All the type names for receivers are encoded in the full key. The key
+// represents the fully qualified package and type name used at link time.
+type packageEscapeFacts struct {
+	Funcs map[string][]Escape
+}
+
+// AFact implements analysis.Fact.AFact.
+func (*packageEscapeFacts) AFact() {}
+
+// Escape is a single escape instances.
+type Escape struct {
+	Reason   string
+	LocalPos token.Pos
+	Resolved LinePosition
```

Comment:
Optional: this field can be removed and replaced with LocalPos.Filename and LocalPos.Line, but it might be nice to keep for convenience.

Reply:
Nope -- see other comment. LocalPos does not have this information.